Mathlib/CategoryTheory/Limits/Shapes/Countable.lean

```diff
@@ -0,0 +1,147 @@
+/-
+Copyright (c) 2023 Dagur Asgeirsson. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Dagur Asgeirsson
+-/
+import Mathlib.CategoryTheory.Limits.Final
+import Mathlib.CategoryTheory.Limits.Shapes.FiniteLimits
+import Mathlib.CategoryTheory.Countable
+import Mathlib.Data.Countable.Defs
+/-!
+# Countable limits and colimits
+
+A typeclass for categories with all countable (co)limits.
+
+We also prove that all cofiltered limits over countable preorders are isomorphic to sequential
+limits, see `sequentialFunctor_initial`.
+
+## Projects
+
+* There is a series of `proof_wanted` at the bottom of this file, implying that all cofiltered
+  limits over countable categories are isomorphic to sequential limits.
+
+* Prove the dual result for filtered colimits.
+
+-/
+
+open CategoryTheory Opposite
+
+variable (C : Type*) [Category C] (J : Type*) [Countable J]
+
+namespace CategoryTheory.Limits
+
+/--
+A category has all countable limits if every functor `J ⥤ C` with a `CountableCategory J`
+instance and `J : Type` has a limit.
+-/
+class HasCountableLimits : Prop where
+  /-- `C` has all limits over any type `J` whose objects and morphisms lie in the same universe
+  and which has countably many objects and morphisms-/
+  out (J : Type) [𝒥 : SmallCategory J] [@CountableCategory J 𝒥] : @HasLimitsOfShape J 𝒥 C _
+
+instance (priority := 100) hasFiniteLimits_of_hasCountableLimits [HasCountableLimits C] :
+    HasFiniteLimits C where
+  out J := HasCountableLimits.out J
+
+instance (priority := 100) hasCountableLimits_of_hasLimits [HasLimits C] :
+    HasCountableLimits C where
+  out := inferInstance
+
```

Comment:
There should be a universe-polymorphic version of `HasCountableLimits.out` using `CountableCategory .objAsTypeEquiv`